node/blockRequestsQueue.go

```diff
@@ -21,38 +21,36 @@ import (
 	"github.com/alvalor/alvalor-go/types"
 )
 
-type transactionRequestsQueue interface {
+type blockRequestsQueue interface {
 	getData() map[string][]types.Hash
 }
 
-type simpleTransactionRequestsQueue struct {
-	requests map[types.Hash][]string
+type simpleBlockRequestsQueue struct {
+	requests []blockRequestMessage
 }
 
-func newTransactionRequestsQueue(requests map[types.Hash][]string) simpleTransactionRequestsQueue {
-	return simpleTransactionRequestsQueue{requests: requests}
+func newBlockRequestsQueue(requests []blockRequestMessage) simpleBlockRequestsQueue {
+	return simpleBlockRequestsQueue{requests: requests}
 }
 
-func (q *simpleTransactionRequestsQueue) getData() map[string][]types.Hash {
+func (q *simpleBlockRequestsQueue) getData() map[string][]types.Hash {
 	data := q.transformToOutgoingRequests()
 	return data
 }
 
 // tranforms map of [tx hashes -> peers addr which have it] to the map of [peer addr -> tx hashes to request]
-func (q *simpleTransactionRequestsQueue) transformToOutgoingRequests() map[string][]types.Hash {
+func (q *simpleBlockRequestsQueue) transformToOutgoingRequests() map[string][]types.Hash {
 	result := make(map[string][]types.Hash)
```

Comment:
Building this kind of map for each request seems an awful lot of work. I would imagine that when we have a request for a transaction, we will simply randomly select a peer from the peers that have the transaction and have the lowest amount of pending requests. Then, we start a timeout timer and forget about it. All we need to do is stop the timer when we receive the transaction, or register a failure for the peer and request it from another when the timer is up.

Reply:
This work is needed to understand how to evenly distribute the load between peers that have these block headers. Selecting randomly from the peers that have transactions and have the lowest amount of pending request is already not random. Another question - how exactly do we figure out the number of pending requests?

Reply:
The point isn't that they should be selected randomly. The point is that we should select the peers that have the lowest number of pending requests. This will constantly give more requests to peers who respond faster and ensure that we optimally distribute over the available peers. Of course, you would need to implement a way to keep track of the requests on each peer, but there is already a data model that keeps track of peer state (it's even called that way). Evenly distributing each batch of requests over peers will not actually create an even distribution, as there is no way to avoid variance, which will create less than optimal distribution of requests over time.

Reply:
Agree with that. On the other hand, if there are no pending requests, you don't really know how to distribute the load. Even distribution still makes sense to me in this case.